### PR TITLE
Add an immutable array type

### PR DESCRIPTION
**Pull request by odersky**
So far, it's called `scala.IArray[T]`. I am not super happy with the name but could not come up with anything better yet.

It's an opaque alias of `Array[T]` that only exposes the `apply` and `length` methods.

Still missing: Wrappers that add all sequence ops to IArrays. We should upgrade to the new collections before we do this.

### CODE REVIEW

**Comment by odersky** [comment]
Btw, the first lines of `IArray.scala` af2a0e6 are a great showcase how opaque types, implied instances, extension methods and inline can work together to give something beautiful.

File: tests/run/iarrays.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import reflect.ClassTag
+object Test extends App {
+
+  val xs = IArray(1, 2, 3)
+
+  def f[T](ys: IArray[T]) = {
```
Comment (julienrf) [inline tests/run/iarrays.scala]: What type does this method erase to? `f(ys: Array[AnyRef])`?

Comment (odersky) [reply tests/run/iarrays.scala]: No, it's `Object`. Just what `Array[T]` erases to.

**Comment by odersky** [comment]
I tried to make `IArray` covariant, using
```scala
opaque type IArray[+T] = Array[_ <: T]
```
Unfortunately, that runs into problems when it comes to class tags. I get:
```
[error] 30 |  def concat[T: ClassTag](xss: IArray[T]*): IArray[T] = Array.concat(xss: _*)
[error]    |                                                                             ^
[error]    |                              No ClassTag available for Array[_ <: T]#T'
[error]    |                              
[error]    |                              where:    T  is a type in method concat
[error]    |                                        T' is a type in class Array
[error] one error found
```
Indeed, there's no way to _create_ an `Array[_ <: T]`. There are probably ways around this, using casts. But it makes me wary to try this as the official definition of `IArray`. `IArray[T]`  should just work out of the box, in exactly the same way `Array[T]` does. But, `Array[T]` and `Array[_ <: T]` are not equivalent in all contexts.

**Comment by sjrd** [comment]
That shouldn't matter outside of the companion of `IArray`, though. Outside, the only thing that can be seen is `IArray[T]`, not any `Array[_ <: T]`. We might require a few hoops inside the definition of `IArray`, but none of that is going to leak outside.

**Comment by odersky** [comment]
@sjrd Right. After adding some casts in iArray's implementation, I could make them covariant, after all. :+1:

File: library/src-bootstrapped/scala/IArray.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package scala
+import reflect.ClassTag
+
+/** It would be nice it IArray could be covariant, but unfortunately that's not
+ *  possible. The problem is the preculiar behavior of `Array[Any]`.
```
Comment (julienrf) [inline library/src-bootstrapped/scala/IArray.scala]: We can now remove that comment :)

Comment (odersky) [reply library/src-bootstrapped/scala/IArray.scala]: Oh, right.

File: library/src-bootstrapped/scala/IArray.scala [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+ *  but it cannot be updated. Unlike regular arrays, immutable arrays are covariant.
+ */
+opaque type IArray[+T] = Array[_ <: T]
+
+object IArray {
+
+  /** Defines extension methods for immutable arrays */
+  implied arrayOps {
+
+    /** The selection operatiom on an immutable array.
```
Comment (julienrf) [inline library/src-bootstrapped/scala/IArray.scala]:
```suggestion
    /** The selection operation on an immutable array.
```

File: tests/pos/covaraint-opqaue.scala [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import annotation.unchecked.uncheckedVariance
```
Comment (julienrf) [inline tests/pos/covaraint-opqaue.scala]: Typo in the file name: “covar*ia*nt-op*aq*ue.scala”

File: tests/run/iarrays.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  val as: IArray[Any] = IArray(1, "hello")
+  assert(as(as.length - 1) == "hello")
+  assert(reduce(as, 0, (x, y) => x.toString ++ y.toString) == "01hello")
+
+  // Check that representation of IArray and Array is the same
+  val bs: IArray[Double] = IArray(1.0, 2.0)
+  val cs: Array[Double] = bs.asInstanceOf[Array[Double]]
+  cs(1) = 3.0
+  assert(bs(1) == 3.0)
```
Comment (julienrf) [inline tests/run/iarrays.scala]: Can we also check that `apply` and `update` don’t box elements?

Comment (odersky) [reply tests/run/iarrays.scala]: That's harder. I believe there are ways to test it, but I don't know them. The fact that `apply` is an inline method means it is literally replaced by its right-hand side, so by the rules of inlining there can't be any boxing. I verified the code after erasure by looking at it to make sure that's the case.

**Comment by odersky** [comment]
Community build fails with:
```
Fetching latest Dotty nightly version...
Latest Dotty nightly build version: 0.14.0-bin-20190221-daf4ce9-NIGHTLY
[info] Set current project to squants (in build file:/tmp/3/community-build/community-projects/squants/)
[info] Forcing Scala version to 0.14.0-bin-SNAPSHOT on all projects.
[info] Reapplying settings...
[info] Set current project to squants (in build file:/tmp/3/community-build/community-projects/squants/)
[info] Updating squantsJVM...
[error] Unable to find credentials for [Sonatype Nexus Repository Manager @ oss.sonatype.org].
[warn] CLIENT ERROR: 401 Unauthorized. Check your resolvers username and password.
```
/cc @OlivierBlanvillain

**Comment by OlivierBlanvillain** [comment]
Are the changes to the community build submodules in c7b0651448dce54a7f16bc891b5fe7a3477373fd intentional?

**Comment by odersky** [comment]
@OlivierBlanvillain  No, I discovered my error: pushed wrong submodule hashes.

**Comment by odersky** [comment]
Going to merge this now, to avoid further breakage by fast-moving infrastructure changes.

**Pull request merged**